jerry-core/ecma/operations/ecma-objects.c

```diff
@@ -1977,6 +1977,51 @@ ecma_op_object_has_instance (ecma_object_t *obj_p, /**< the object */
   return ecma_raise_type_error (ECMA_ERR_MSG ("Expected a function object."));
 } /* ecma_op_object_has_instance */
 
+/**
+ * General [[GetPrototypeOf]] abstract operation
+ *
+ * Note: returned valid object must be freed.
+ *
+ * @return ecma_object_t * - prototype of the input object.
+ *         ECMA_OBJECT_POINTER_ERROR - error reported during Proxy resolve.
+ *         NULL - the input object does not have a prototype.
+ */
+ecma_object_t *
+ecma_op_object_get_prototype_of (ecma_object_t *obj_p) /**< input object */
+{
+  JERRY_ASSERT (obj_p != NULL);
+
+#if ENABLED (JERRY_BUILTIN_PROXY)
+  if (ECMA_OBJECT_IS_PROXY (obj_p))
+  {
+    ecma_value_t proto = ecma_proxy_object_get_prototype_of (obj_p);
+
+    if (ECMA_IS_VALUE_ERROR (proto))
+    {
+      return ECMA_OBJECT_POINTER_ERROR;
+    }
+    else if (ecma_is_value_null (proto))
+    {
+      return NULL;
+    }
+
+    JERRY_ASSERT (ecma_is_value_object (proto));
+
+    return ecma_get_object_from_value (proto);
+  }
+  else
+#endif /* ENABLED (JERRY_BUILTIN_PROXY) */
+  {
+    ecma_object_t *proto = ECMA_GET_POINTER (ecma_object_t, ecma_op_ordinary_object_get_prototype_of (obj_p));
```

Comment:
I see a double null check here, since `ECMA_GET_POINTER ` also checks for JMEM_CP_NULL. I suggest:

```c
jmem_cpointer_t proto_cp = ecma_op_ordinary_object_get_prototype_of (func_obj_p);

if (proto_cp == JMEM_CP_NULL)
{   
  return NULL;
}

ecma_object_t *proto_p = ECMA_GET_NON_NULL_POINTER (ecma_object_t, proto_cp);
ecma_ref_object (proto_p );
return proto_p;
```